### PR TITLE
[FLINK-7748][network] properly use the TaskEventDispatcher for subscribing to events

### PR DESCRIPTION
**Pull request by NicoK**
## What is the purpose of the change

`ResultPartitionWriter` currently implements the `EventListener` interface and is used for event registration, although event publishing is already handled via the `TaskEventDispatcher`.
Instead of using two different places, this should be unified by using `TaskEventDispatcher` only which is done by this PR.

~Please note that this PR builds upon #4759.~

## Brief change log

- make `TaskEventDispatcher` more generic to register result partitions via `ResultPartitionID` with one `TaskEventHandler` per partition (handled by `TaskEventDispatcher`, not inside `ResultPartitionWriter`)
- remove the `EventListener<TaskEvent>` implementation from `ResultPartitionWriter`
- add the `TaskEventDispatcher` to the `Environment` information for a task to be able to work with it (only `IterationHeadTask` is using this as of now)
- adapt all places to use `TaskEventDispatcher` instead of `ResultPartitionWriter`

## Verifying this change

This change added tests and can be verified as follows:

- a new `TaskEventDispatcherTest` verifies that `TaskEventDispatcher` works as expected
- indirectly, tests under `org.apache.flink.runtime.iterative` may verify the use of the `IterationHeadTask` which is the only user so far

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

**Comment by pnowojski** [comment]
Could you resolve conflicts before I start reviewing?

**Comment by NicoK** [comment]
alright - didn't even need to do anything here since git auto-merged them :)

File: flink-runtime/src/main/java/org/apache/flink/runtime/iterative/task/IterationHeadTask.java [hunk 1]
```diff
@@ -223,8 +225,10 @@ private void readInitialSolutionSet(JoinHashMap<X> solutionSet, MutableObjectIte
 
 	private SuperstepBarrier initSuperstepBarrier() {
 		SuperstepBarrier barrier = new SuperstepBarrier(getUserCodeClassLoader());
-		this.toSync.subscribeToEvent(barrier, AllWorkersDoneEvent.class);
-		this.toSync.subscribeToEvent(barrier, TerminationEvent.class);
+		TaskEventDispatcher taskEventDispatcher = getEnvironment().getTaskEventDispatcher();
+		ResultPartitionID partitionId = this.toSync.getPartitionId();
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/iterative/task/IterationHeadTask.java]: nit: remove `this`

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriter.java [hunk 2]
```diff
@@ -34,12 +31,10 @@
  * The {@link ResultPartitionWriter} is the runtime API for producing results. It
  * supports two kinds of data to be sent: buffers and events.
  */
-public class ResultPartitionWriter implements EventListener<TaskEvent> {
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriter.java]: I do not understand why this class was previously implementing `EventListener`. I couldn't find any context in which it was used that way.

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java [hunk 3]
```diff
@@ -671,7 +671,7 @@ else if (current == ExecutionState.CANCELING) {
 				jobConfiguration, taskConfiguration, userCodeClassLoader,
 				memoryManager, ioManager, broadcastVariableManager,
 				accumulatorRegistry, kvStateRegistry, inputSplitProvider,
-				distributedCacheEntries, writers, inputGates,
+				distributedCacheEntries, writers, inputGates, network.getTaskEventDispatcher(),
```
Comment (pnowojski) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java]: wow. Could you reformat this one with one parameter per line?

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/TaskEventDispatcherTest.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+		expectedException.expectMessage("not registered at task event dispatcher");
+
+		//noinspection unchecked
+		ted.subscribeToEvent(new ResultPartitionID(), mock(EventListener.class), TaskEvent.class);
+	}
+
+	/**
+	 * Tests {@link TaskEventDispatcher#publish(ResultPartitionID, TaskEvent)} and {@link TaskEventDispatcher#subscribeToEvent(ResultPartitionID, EventListener, Class)} methods.
+	 */
+	@Test
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/TaskEventDispatcherTest.java]: Guess what? :)

```
private static OneShotEventListener extends EventListener<TaskEvent> {
  private final TaskEvent expected;
  private boolean fired = false;
  public OneShotEventListener(TaskEvent expected) {
    ...
  }
  
  public void onEvent(TaskEvent actual) {
     checkState(!fired);
     fired = true;
     checkArgument(actual == expected);
  }
}

private static ZeroShotEventListener extends EventListener<TaskEvent> {
  public void onEvent(TaskEvent actual) {
     throw new IllegalStateException();
  }
}
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/TaskEventDispatcherTest.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+		assertFalse(ted.publish(partitionId1, event1));
+
+		ted.registerPartition(partitionId1);
+		ted.registerPartition(partitionId2);
+
+		// no event listener subscribed yet, but the event is forwarded to a TaskEventHandler
+		assertTrue(ted.publish(partitionId1, event1));
+
+		//noinspection unchecked
+		EventListener<TaskEvent> eventListener1a = mock(EventListener.class);
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/TaskEventDispatcherTest.java]:
```
 = new OneShotEventListener(event1);
 = new ZeroShotEventListener();
 ...
```

File: flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java [hunk 6]
```diff
@@ -325,9 +325,10 @@ public void testRequestBackoffConfiguration() throws Exception {
 		int initialBackoff = 137;
 		int maxBackoff = 1001;
 
+		TaskEventDispatcher taskEventDispatcher = new TaskEventDispatcher();
 		NetworkEnvironment netEnv = mock(NetworkEnvironment.class);
 		when(netEnv.getResultPartitionManager()).thenReturn(new ResultPartitionManager());
-		when(netEnv.getTaskEventDispatcher()).thenReturn(new TaskEventDispatcher());
+		when(netEnv.getTaskEventDispatcher()).thenReturn(taskEventDispatcher);
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateTest.java]: what is the purpose of this change?

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DummyEnvironment.java [hunk 7]
```diff
@@ -190,4 +191,8 @@ public InputGate getInputGate(int index) {
 		return null;
 	}
 
+	@Override
+	public TaskEventDispatcher getTaskEventDispatcher() {
+		return null;
```
Comment (pnowojski) [inline flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DummyEnvironment.java]: `throw new UnsuportedOperationException()` ?

File: flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriter.java (hunk 2 above)
Comment (NicoK) [reply flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/writer/ResultPartitionWriter.java]: a quick search reveals `NetworkEnvironment.java` calling `taskEventDispatcher.registerWriterForIncomingTaskEvents(writer.getPartitionId(), writer);` and therefore `TaskEventDispatcher` itself but there, the `EventListener` interface was actually only used in `TaskEventDispatcher#publish()` which is somewhat strange - agreed

File: flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DummyEnvironment.java (hunk 7 above)
Comment (NicoK) [reply flink-runtime/src/test/java/org/apache/flink/runtime/operators/testutils/DummyEnvironment.java]: ok, let's try this out

**Comment by zentol** [comment]
merging.
